HPCC-18780 Generate projected as well as expected disk meta functions

richardkchapman opened the pull request:
Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:
Compiled
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

richardkchapman committed:
HPCC-18780 Generate projected as well as expected disk meta functions

Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

richardkchapman commented:
@ghalliday Please review

HPCCSmoketest commented:
Automated Smoketest: :x:
Sha: ca87909befe350dbaeb6f75698bac84a31fb7b22
Build: success
ECL Watch:  Rebuilding Site
Empty table

Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |116 |116 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |727 |696 |31 |
|test (thor) |640 |608 |32 |
|test (roxie) |755 |682 |73 |


[Errors:](https://gist.github.com/60ee859d70099321c9cfb9a595fec68f)
- test
  - hthor
    - 63\. [bug5236 (multiPart=false)](https://gist.github.com/HPCCSmoketest/60ee859d70099321c9cfb9a595fec68f#file-bug5236-multipart-false-w20171121-100858-3-log), 64\. [bug5236 (multiPart=true)](https://gist.github.com/HPCCSmoketest/60ee859d70099321c9cfb9a595fec68f#file-bug5236-multipart-true-w20171121-100859-log), 65\. [bug5236 (multiPart=true, useLocal=true)](https://gist.github.com/HPCCSmoketest/60ee859d70099321c9cfb9a595fec68f#file-bug5236-multipart-true,-uselocal-true-w20171121-100859-4-log), 150\. [dfsj](https://gist.github.com/HPCCSmoketest/60ee859d70099321c9cfb9a595fec68f#file-dfsj-w20171121-100916-6-log), 182\. [dict_index](https://gist.github.com/HPCCSmoketest/60ee859d70099321c9cfb9a595fec68f#file-dict_index-w20171121-100925-4-log), 207\. [dynamicoptflag](https://gist.github.com/HPCCSmoketest/60ee859d70099321c9cfb9a595fec68f#file-dynamicoptflag-w20171121-100931-log), 293\. [indexread2 (multiPart=false)](https://gist.github.com/HPCCSmoketest/60ee859d70099321c9cfb9a595fec68f#file-indexread2-multipart-false-w20171121-100954-1-log), 294\. [indexread2 (multiPart=true)](https://gist.github.com/HPCCSmoketest/60ee859d70099321c9cfb9a595fec68f#file-indexread2-multipart-true-w20171121-100955-5-log), 295\. [indexread2 (multiPart=true, useLocal=true)](https://gist.github.com/HPCCSmoketest/60ee859d70099321c9cfb9a595fec68f#file-indexread2-multipart-true,-uselocal-true-w20171121-100955-log), 345\. [keyed_denormalize (multiPart=false)](https://gist.github.com/HPCCSmoketest/60ee859d70099321c9cfb9a595fec68f#file-keyed_denormalize-multipart-false-w20171121-101006-2-log), 346\. [keyed_denormalize (multiPart=true)](https://gist.github.com/HPCCSmoketest/60ee859d70099321c9cfb9a595fec68f#file-keyed_denormalize-multipart-true-w20171121-101006-4-log), 347\. [keyed_denormalize (multiPart=true, useLocal=true)](https://gist.github.com/HPCCSmoketest/60ee859d70099321c9cfb9a595fec68f#file-keyed_denormalize-multipart-true,-uselocal-true-w20171121-101007-log), 351\. [keyed_join (multiPart=false)](https://gist.github.com/HPCCSmoketest/60ee859d70099321c9cfb9a595fec68f#file-keyed_join-multipart-false-w20171121-101008-2-log), 352\. [keyed_join (multiPart=true)](https://gist.github.com/HPCCSmoketest/60ee859d70099321c9cfb9a595fec68f#file-keyed_join-multipart-true-w20171121-101008-6-log), 353\. [keyed_join (multiPart=true, useLocal=true)](https://gist.github.com/HPCCSmoketest/60ee859d70099321c9cfb9a595fec68f#file-keyed_join-multipart-true,-uselocal-true-w20171121-101008-4-log), 354\. [keyed_join2 (multiPart=false)](https://gist.github.com/HPCCSmoketest/60ee859d70099321c9cfb9a595fec68f#file-keyed_join2-multipart-false-w20171121-101009-2-log), 355\. [keyed_join2 (multiPart=true)](https://gist.github.com/HPCCSmoketest/60ee859d70099321c9cfb9a595fec68f#file-keyed_join2-multipart-true-w20171121-101009-log), 356\. [keyed_join2 (multiPart=true, useLocal=true)](https://gist.github.com/HPCCSmoketest/60ee859d70099321c9cfb9a595fec68f#file-keyed_join2-multipart-true,-uselocal-true-w20171121-101009-5-log), 357\. [keye
 ... ( comment is too long, 40 lines, 20843 bytes, truncated) ... 
8-5-log), 740\. [when8](https://gist.github.com/HPCCSmoketest/60ee859d70099321c9cfb9a595fec68f#file-when8-w20171121-102309-log), 741\. [when9](https://gist.github.com/HPCCSmoketest/60ee859d70099321c9cfb9a595fec68f#file-when9-w20171121-102309-1-log), 

```

HPCCSmoketest commented:
Automated Smoketest: :x:
Sha: 33156e60f74425e8b1e39c49a163351ff7b65681
Build: success
ECL Watch:  Rebuilding Site

| errors | 
|---|
|5 |



eclwatch/LFDetailsWidget.js: line 126, col 33, targetRow is not defined.
eclwatch/LFDetailsWidget.js: line 138, col 40, targetRow is not defined.
eclwatch/LFDetailsWidget.js: line 139, col 29, targetRow is not defined.
eclwatch/LFDetailsWidget.js: line 140, col 29, targetRow is not defined.
eclwatch/LFDetailsWidget.js: line 142, col 77, targetRow is not defined.
Error(s): 

Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |116 |116 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |727 |727 |0 |
|test (thor) |640 |637 |3 |
|test (roxie) |755 |755 |0 |


[Errors:](https://gist.github.com/4a8b10a70a7cb8dd1a28a3ed48accbe3)
- test
  - thor
    - 558\. [steplimit2 (multiPart=false)](https://gist.github.com/HPCCSmoketest/4a8b10a70a7cb8dd1a28a3ed48accbe3#file-steplimit2-multipart-false-steplimit2-ecl-log), 559\. [steplimit2 (multiPart=true)](https://gist.github.com/HPCCSmoketest/4a8b10a70a7cb8dd1a28a3ed48accbe3#file-steplimit2-multipart-true-steplimit2-ecl-log), 626\. [wuread3](https://gist.github.com/HPCCSmoketest/4a8b10a70a7cb8dd1a28a3ed48accbe3#file-wuread3-wuread3-ecl-log), 

HPCC Stop: OK
HPCC Uninstall: OK

richardkchapman committed:
HPCC-18786 Thor workunit read from disk may call random method

When a diskread activity is created in place of a workunit read, one of the
virtual calls to the helper is made using the wrong virtual table (it's using
the original activity helper pointer, but cast to a IHThorDiskReadBaseArg,
which it is not).

Revealed when VMT of IHThorDiskReadBaseArg changed, and the random function
being called started to be a destructor.

Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

richardkchapman commented:
@jakesmith Please note/review the lurking thor bug that this change revealed. You may prefer to fix it differently (not sure why you don't overwrite basehelper in the container before construction for example).

There is similar code on the master side, though I don't think it actually makes any calls through the helper before it is updated, it's potentially an accident waiting to happen still.

I don't know what if any ill effects this might be causing in pre 7.0 versions and whether worth backporting the fix to them.

HPCCSmoketest commented:
Automated Smoketest: :x:
Sha: 49329e76c1594c9f2c24dab7fdf9ffe2e94a2205
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |116 |116 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |727 |727 |0 |
|test (thor) |640 |638 |2 |
|test (roxie) |755 |755 |0 |


[Errors:](https://gist.github.com/d0b9c43ccd3d51a0b77399d0c2ef63f3)
- test
  - thor
    - 558\. [steplimit2 (multiPart=false)](https://gist.github.com/HPCCSmoketest/d0b9c43ccd3d51a0b77399d0c2ef63f3#file-steplimit2-multipart-false-steplimit2-ecl-log), 559\. [steplimit2 (multiPart=true)](https://gist.github.com/HPCCSmoketest/d0b9c43ccd3d51a0b77399d0c2ef63f3#file-steplimit2-multipart-true-steplimit2-ecl-log), 

HPCC Stop: OK
HPCC Uninstall: OK

richardkchapman commented on thorlcr/activities/thdiskbaseslave.cpp:
```diff
@@ -202,8 +202,10 @@ const char * CDiskPartHandlerBase::queryLogicalFilename(const void * row)
+    if (_helper)
+        baseHelper.set(_helper);
     helper = (IHThorDiskReadBaseArg *)queryHelper();
     reInit = 0 != (helper->getFlags() & (TDXvarfilename|TDXdynamicfilename));
```
Prior to my change, this might act on the original WorkUnitReadArg

jakesmith commented:
@richardkchapman 
>not sure why you don't overwrite basehelper in the container before construction for example

Nor am I. I think the container itself should be converted, i.e. the detection should happen much earlier, so there's no potential confusion downstream in either master or slave.

>There is similar code on the master side, though I don't think it actually makes any calls through the helper before it is updated, it's potentially an accident waiting to happen still.

I've checked and it's safe, but as you a potential problem to trip over in the future.

However, I think can go with this change and I'll open a new JIRA to look at refactoring the conversion so it happens earlier.

@ghalliday - can you review the rest ?

HPCCSmoketest commented:
Automated Smoketest: :x:
Sha: 49329e76c1594c9f2c24dab7fdf9ffe2e94a2205
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |116 |116 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |727 |727 |0 |
|test (thor) |640 |638 |2 |
|test (roxie) |755 |755 |0 |


[Errors:](https://gist.github.com/c95b7374e530bee0927705f4b76a58b8)
- test
  - thor
    - 558\. [steplimit2 (multiPart=false)](https://gist.github.com/HPCCSmoketest/c95b7374e530bee0927705f4b76a58b8#file-steplimit2-multipart-false-steplimit2-ecl-log), 559\. [steplimit2 (multiPart=true)](https://gist.github.com/HPCCSmoketest/c95b7374e530bee0927705f4b76a58b8#file-steplimit2-multipart-true-steplimit2-ecl-log), 

HPCC Stop: OK
HPCC Uninstall: OK
